contributors/design-proposals/sig-cli/kubectl_plan.md

```diff
@@ -0,0 +1,164 @@
+by SIG cli to keep up with changes made on the server side.  Server side
+changes frequently break the cli creation commands, requiring the SIG cli
+maintainers to track down the cause and sometimes implement a fix.
+
+This document proposes introducing an optional `plan` subresource which converts the
+commandline arguments into an object on the server side through a subresource and
+simplified representation of the object.  The object is not stored as part of the request,
+instead the client must do so through a separate request.
+
+## Goals
+
+- Provide simple cli commands that reflect how the system is designed
+- Enable API extensions to expose cli commands without downloading plugins 
+- Allow implementers of resource APIs to define cli experience without
+  creating PRs in kubectl
+- Reduce sustaining engineering burden on SIG to maintain the correctness
+  of the commands
+
+## Plan CLI
+
+Kubectl will expose the new `plan` and `plan-list` top-level commands.
+
+When a user runs `kubectl plan -h` it will print a help message
+describing the command, and prompt the user to run `plan-list` to view
+the available subcommands.
+
+**Note:** Alternatively, we could just make the help message itself dynamic instead
+of introducing a `plan-list` command.  We should make this consistent with whatever
+we end up doing for `kubectl get`.
+
+`plan` and `plan-list` will use the discovery API to identify all resources with a `plan`
+subcommand.  It will lookup the request Model from the openapi, and check that
+the model contains a field called `PlanArguments` verify that the
+subresource declares that it implements the interface, and is not the result of a
+naming collision.  This will also make it easy to add additional fields to the request
+that may not map directly to cli flags.
+
+When a user runs `kubectl plan <resource> -h`, kubectl will parse the help message
+for the available flags from the fields in `Flags`.
+The flag descriptions will be pulled from the field description in the openapi model,
+and the names will convert the *camelCase* field names to *dash-delimited* flag names.
+The name and namespace of the object will be derived from the url path as they are
+done for `get`.
+
+TODO: Find a better name than *Flags* for the field name
+
+## Plan subresource API
+
+Resources can *optionally* expose a `plan` subresource the contains a field named
+*Flags*.  The kubectl will expose this subresource directly as a subcommand by
+parsing the *Flags* argument into commandline flags based off the openapi model.
+
+```go
+type DeploymentPlanRequest struct {
+    Flags DeploymentArguments
+}
+
+// deploymentArguments contains the values converted from commandline arguments
+type DeploymentArguments struct {
+    // images is the images to run as separate containers in a Pod
+    Images []string
+    
+    // replicas is the number of replicas to set
+    Replicas int32
+}
+```
+
+The subresource will expose the API as a POST operation.
```

Comment:
why POST ? this command doesn't not actually create, also what about the endpoints of this new subresource ? do we need some new endpoints of apiserver, that means it also need some server side implementation to support this new command ?

Reply:
POST means "do something" in HTTP.  It doesn't mean create.